core/src/test/java/io/atomix/core/AtomixTest.java

```diff
@@ -82,4 +89,80 @@ public void testScaleDown() throws Exception {
     instances.get(1).stop().join();
     instances.get(2).stop().join();
   }
+
+  /**
+   * Tests a client joining and leaving the cluster.
+   */
+  @Test
+  public void testClientJoinLeave() throws Exception {
+    List<CompletableFuture<Atomix>> futures = new ArrayList<>();
+    futures.add(startAtomix(Node.Type.DATA, 1, 1, 2, 3));
+    futures.add(startAtomix(Node.Type.DATA, 2, 1, 2, 3));
+    futures.add(startAtomix(Node.Type.DATA, 3, 1, 2, 3));
+    Futures.allOf(futures).join();
+
+    TestClusterEventListener dataListener = new TestClusterEventListener();
+    instances.get(0).clusterService().addListener(dataListener);
+
+    Atomix client1 = startAtomix(Node.Type.CLIENT, 4, 1, 2, 3).join();
+
+    // client1 added to data node
+    ClusterEvent event1 = dataListener.event();
+    assertEquals(ClusterEvent.Type.NODE_ADDED, event1.type());
+    event1 = dataListener.event();
+    assertEquals(ClusterEvent.Type.NODE_ACTIVATED, event1.type());
+
+    TestClusterEventListener clientListener = new TestClusterEventListener();
+    client1.clusterService().addListener(clientListener);
+
+    Atomix client2 = startAtomix(Node.Type.CLIENT, 5, 1, 2, 3).join();
+
+    // client2 added to data node
+    ClusterEvent event2 = dataListener.event();
+    assertEquals(ClusterEvent.Type.NODE_ADDED, event2.type());
+    event2 = dataListener.event();
+    assertEquals(ClusterEvent.Type.NODE_ACTIVATED, event2.type());
+
+    // client2 added to client node
+    event1 = clientListener.event();
+    assertEquals(ClusterEvent.Type.NODE_ADDED, event1.type());
+    event1 = clientListener.event();
+    assertEquals(ClusterEvent.Type.NODE_ACTIVATED, event1.type());
+
+    client2.stop().join();
+
+    // client2 removed from data node
+    event1 = dataListener.event();
+    assertEquals(ClusterEvent.Type.NODE_DEACTIVATED, event1.type());
+    event1 = dataListener.event();
+    assertEquals(ClusterEvent.Type.NODE_REMOVED, event1.type());
+
+    // client2 removed from client node
+    event1 = clientListener.event();
+    assertEquals(ClusterEvent.Type.NODE_DEACTIVATED, event1.type());
+    event1 = clientListener.event();
+    assertEquals(ClusterEvent.Type.NODE_REMOVED, event1.type());
+  }
+
+  private static class TestClusterEventListener implements ClusterEventListener {
+
+    private final BlockingQueue<ClusterEvent> queue = new ArrayBlockingQueue<>(1);
+
+    @Override
+    public void onEvent(ClusterEvent event) {
+      try {
+        queue.put(event);
+      } catch (InterruptedException e) {
+        Throwables.propagate(e);
```

Comment:
Why propagate and not simply Thread.currentThread().interrupt()

Reply:
Just catching a checked exception and rethrowing it as an unchecked exception

Reply:
I think we should restore the interrupted status

Reply:
https://www.ibm.com/developerworks/library/j-jtp05236/index.html